### ISSUE TITLE
Cilium assigns additional IP to cilium_host upon restart for IPAMCRD mode

### ISSUE THREAD

**Issue by houminz** [cilium/cilium#20879]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Hi，I came across the same problem as #16672 and #16307 with cilium version v1.10.4 

In my case, IPAM is crd mode. upon Cilium restart, it was very likely that cilium_host was assigned an additional IP. 

For example, pod cidr 10.24.0.0/26 is allocated for one node, and originally `cilium_host` ip is `10.24.0.55`

```
apiVersion: v1
kind: Node
metadata:
  annotations:
    io.cilium.network.ipv4-cilium-host: 10.24.0.24
    io.cilium.network.ipv4-health-ip: 10.24.0.20
    io.cilium.network.ipv4-pod-cidr: 10.24.0.0/26
```
After I rebuild the `cilium` pod, the cilium_host was assigned an additional IP.

```
4: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether c2:35:33:f2:ab:00 brd ff:ff:ff:ff:ff:ff
    inet 10.24.0.55/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet 10.24.0.24/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet6 fe80::c035:33ff:fef2:ab00/64 scope link
       valid_lft forever preferred_lft forever
```

As I try to read the code and cilium log, we can see that 2 parts of router ip restoration both get `10.24.0.55`, i.e. `routerFromK8s` and `routerFromFs` are both `10.24.0.55`.

Howerver, in crd mode, cidr is nil, thus leading the log `restoration does not belong in the Pod CIDR of the node`

```
level=info msg="Restored router IPs from node information" ipv4=10.24.0.55 ipv6="<nil>" subsys=k8s
level=info msg="k8s mode: Allowing localhost to reach local endpoints" subsys=daemon
level=info msg="Enabling k8s event listener" subsys=k8s-watcher
level=info msg="Removing stale endpoint interfaces" subsys=daemon
level=info msg="Skipping kvstore configuration" subsys=daemon
level=info msg="Restored router address from node_config" file=/var/run/cilium/state/globals/node_config.h ipv4=10.24.0.55 ipv6="<nil>" subsys=node
level=info msg="Initializing node addressing" subsys=daemon
level=info msg="Initializing CRD-based IPAM" subsys=ipam
level=info msg="Subscribed to CiliumNode custom resource" name=10.0.24.205 subsys=ipam
level=info msg="Creating or updating CiliumNode resource" node=10.0.24.205 subsys=nodediscovery
level=info msg="Waiting until all pre-existing resources related to policy have been received" subsys=k8s-watcher
level=info msg="Waiting for CiliumNode custom resource to become available..." name=10.0.24.205 subsys=ipam
level=info msg="All pre-existing resources related to policy have been received; continuing" subsys=k8s-watcher
level=info msg="Successfully synchronized CiliumNode custom resource" name=10.0.24.205 subsys=ipam
level=info msg="All required IPs are available in CRD-backed allocation pool" available=62 name=10.0.24.205 required=2 subsys=ipam
level=info msg="The router IP (10.24.0.55) considered for restoration does not belong in the Pod CIDR of the node. Discarding old router IP." cidr="<nil>" subsys=node
```

As the code shows, in crd mode, cidr is from `IPv4NativeRoutingCIDR()`

https://github.com/cilium/cilium/pull/16672/files#diff-2ed5415f522d0b72066d62527fb40dbea9bf745f45d5e2d47f7715680f165fcaR278

```go
if ipam := option.Config.IPAMMode(); ipam == ipamOption.IPAMCRD || ipam == ipamOption.IPAMENI || ipam == ipamOption.IPAMAlibabaCloud {
	// The native routing CIDR is the pod CIDR in these IPAM modes.
	cidr = option.Config.IPv4NativeRoutingCIDR()
} else {
	cidr = node.GetIPv4AllocRange()
}
```
But the `IPv4NativeRoutingCIDR()` is only set for IPAMENI/IPAMAzure/IPAMAlibabaCloud, not in IPAMCRD. So for my scenario, cidr is nil.

```go 
if n.conf.IPAMMode() == ipamOption.IPAMENI || n.conf.IPAMMode() == ipamOption.IPAMAzure || n.conf.IPAMMode() == ipamOption.IPAMAlibabaCloud {
	if vpcCIDR := deriveVpcCIDR(n.ownNode); vpcCIDR != nil {
		if nativeCIDR := n.conf.IPv4NativeRoutingCIDR(); nativeCIDR != nil {
			logFields := logrus.Fields{
				"vpc-cidr":                   vpcCIDR.String(),
				option.IPv4NativeRoutingCIDR: nativeCIDR.String(),
			}

			ranges4, _ := ip.CoalesceCIDRs([]*net.IPNet{nativeCIDR.IPNet, vpcCIDR.IPNet})
			if len(ranges4) != 1 {
				log.WithFields(logFields).Fatal("Native routing CIDR does not contain VPC CIDR.")
			} else {
				log.WithFields(logFields).Info("Ignoring autodetected VPC CIDR.")
			}
		} else {
			log.WithFields(logrus.Fields{
				"vpc-cidr": vpcCIDR.String(),
			}).Info("Using autodetected VPC CIDR.")
			n.conf.SetIPv4NativeRoutingCIDR(vpcCIDR)
		}
	} else {
		minimumReached = false
	}
}
```
https://github.com/cilium/cilium/blob/v1.10.4/pkg/ipam/crd.go#L273

So maybe code here should not include condition IPAMCRD, or maybe the usage in my case is not correctly? Any help would be appreciated.

```go
if ipam := option.Config.IPAMMode(); ipam == ipamOption.IPAMCRD || ipam == ipamOption.IPAMENI || ipam == ipamOption.IPAMAlibabaCloud {
	// The native routing CIDR is the pod CIDR in these IPAM modes.
	cidr = option.Config.IPv4NativeRoutingCIDR()
} else {
	cidr = node.GetIPv4AllocRange()
}
```

### Cilium Version

v1.10.4

### Kernel Version

Linux VM-15-171-centos 5.4.119-19-0009.3 #1 SMP Wed Apr 20 22:27:43 CST 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.5-tke.1", GitCommit:"67551df2c894ff13cf63e0d926d05bba8fa2830a", GitTreeState:"clean", BuildDate:"2022-01-05T09:29:37Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22+", GitVersion:"v1.22.5-tke.1", GitCommit:"79b517faeb244730136bef57c694da493a413bf9", GitTreeState:"clean", BuildDate:"2022-04-19T03:56:15Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by SimpCosm** [comment]
sysdump info is here 
[cilium-sysdump-20220812-171433.zip](https://github.com/cilium/cilium/files/9315191/cilium-sysdump-20220812-171433.zip)

**Comment by SimpCosm** [comment]
#17762 fix this issue to Remove old, discarded router IPs from cilium_host
it was backported to v1.10.6 #18088 #18214

#19341 also Consider VPC's secondary CIDRs during cilium_host IP restoration

**Comment by christarazi** [comment]
Moving the discussion from https://github.com/cilium/cilium/pull/17762#issuecomment-1233870638 to here: if `restoredIP` is nil, do you have any ideas on how we might detect the addr family? Would relying on Cilium's configuration help? That raises the question of what happens if both IPv4 and IPv6 are enabled.

**Comment by christarazi** [comment]
Re: https://github.com/cilium/cilium/pull/16672#issuecomment-1211521860

This code https://github.com/cilium/cilium/blob/v1.10.4/pkg/ipam/crd.go#L273 missing `IPAMCRD` in its condition seems like an oversight. It should align with the other conditions which check all the other CRD-backed IPAM modes. Would you be able to open a PR to fix this?

**Comment by SimpCosm** [comment]
> Moving the discussion from [#17762 (comment)](https://github.com/cilium/cilium/pull/17762#issuecomment-1233870638) to here: if `restoredIP` is nil, do you have any ideas on how we might detect the addr family? Would relying on Cilium's configuration help? That raises the question of what happens if both IPv4 and IPv6 are enabled.

Yes, I found this problem is fixed https://github.com/cilium/cilium/pull/20389 here

**Comment by SimpCosm** [comment]
@christarazi sure I can open a PR to fix this by changing the code below, and it do solve the problem in my test. Or maybe another solution is to merge the patch from https://github.com/cilium/cilium/pull/20389

```go
if ipam := option.Config.IPAMMode(); ipam == ipamOption.IPAMCRD || ipam == ipamOption.IPAMENI || ipam == ipamOption.IPAMAlibabaCloud {
	// The native routing CIDR is the pod CIDR in these IPAM modes.
	cidr = option.Config.IPv4NativeRoutingCIDR()
} else {
	cidr = node.GetIPv4AllocRange()
}
```

to 

```go
if ipam := option.Config.IPAMMode(); ipam == ipamOption.IPAMCRD || ipam == ipamOption.IPAMENI || ipam == ipamOption.IPAMAlibabaCloud {
	// The native routing CIDR is the pod CIDR in these IPAM modes.
	cidr = option.Config.IPv4NativeRoutingCIDR()
} else {
	cidr = node.GetIPv4AllocRange()
}
```

**Comment by christarazi** [comment]
@SimpCosm Are you saying that if you run with the following commits

fcd00390c30c6eeffbe2fefa81d5b22e59397297
02fa124f73e44cde4124c9f37325ce66c338aa98 
bcf5f25948827bce437c61eff2efbc41e55662fd

that it resolves your issue? These should all be in 1.10.13.

If so, please upgrade to that version. If that resolves your issue, then please close this issue as well.

**Comment by SimpCosm** [comment]
> 

Yes，I backport this commit to v1.10.4 and it solves my issue. Thanks anyway.

https://github.com/cilium/cilium/commit/fcd00390c30c6eeffbe2fefa81d5b22e59397297
https://github.com/cilium/cilium/commit/02fa124f73e44cde4124c9f37325ce66c338aa98
https://github.com/cilium/cilium/commit/bcf5f25948827bce437c61eff2efbc41e55662fd
